plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Completions.hs

```diff
@@ -0,0 +1,681 @@
+  makeFullDirPath completion' complInfo = do
+    let fullPath = partialFileDir complInfo Posix.</> T.unpack completion'
+    pure $ T.pack fullPath
+
+{- | Takes a path completion info and returns the list of files
+  in the directory that match the path completion info.
+-}
+listFileCompletions :: Recorder (WithPriority Log) -> PathCompletionInfo -> IO [FilePath]
+listFileCompletions recorder complInfo = do
+  let complDir = mkCompletionDirectory complInfo
+  try (evaluate =<< listDirectory complDir) >>= \case
+    Right dirs -> do
+      fixedDirs <-
+        mapM
+          ( \d -> do
+              isDir <- doesDirectoryExist $ mkDirFromCWD complInfo d
+              pure $ if isDir then Posix.addTrailingPathSeparator d else d
+          )
+          dirs
+      pure fixedDirs
+    Left (err :: IOError) -> do
+      logWith recorder Warning $ LogFilePathCompleterIOError complDir err
+      pure []
+
+{- | Returns a list of all (and only) directories in the
+  directory described by path completion info
+-}
+listDirectoryCompletions :: Recorder (WithPriority Log) -> PathCompletionInfo -> IO [FilePath]
+listDirectoryCompletions recorder complInfo = do
+  filepaths <- listFileCompletions recorder complInfo
+  filterM (doesDirectoryExist . mkDirFromCWD complInfo) filepaths
+
+pathCompletionInfoFromCompletionContext :: CabalCompletionContext -> PathCompletionInfo
+pathCompletionInfoFromCompletionContext ctx =
+  PathCompletionInfo
+    { partialFileName = dirNamePrefix
+    , partialFileDir = Posix.addTrailingPathSeparator $ Posix.takeDirectory prefix
+    , cabalFileDir = dir
+    }
+ where
+  prefix = T.unpack $ completionPrefix ctx
+  dirNamePrefix = T.pack $ Posix.takeFileName prefix
+  dir = Posix.takeDirectory $ completionCabalFileDir ctx
+
+{- | Returns the directory, the currently handled cabal file is in.
+
+  We let System.FilePath handle the separator syntax since this is used
+  to query filepaths from the system. See Note [Using correct file path separators].
+-}
+mkCompletionDirectory :: PathCompletionInfo -> FilePath
+mkCompletionDirectory complInfo =
+  FP.addTrailingPathSeparator $
+    cabalFileDir complInfo FP.</> (FP.normalise $ partialFileDir complInfo)
+
+{- | Returns the complete filepath for the given filepath.
+
+  Since this is used for completions we use posix separators here.
+  See Note [Using correct file path separators].
+-}
+mkDirFromCWD :: PathCompletionInfo -> FilePath -> FilePath
+mkDirFromCWD complInfo fp =
+  Posix.addTrailingPathSeparator $
+    mkCompletionDirectory complInfo Posix.</> Posix.normalise fp
+
+{- | Information used to query and build file path/directory completions.
+
+  Note that partialFileName combined with partialFileDir results in
+  the original prefix.
+
+  Example:
+  On the written filepath: @dir1/fi@ the
+  resulting PathCompletionInfo would be:
+
+  @
+    partialFileName = "fi"
+    partialFileDir  = "dir1/dir2/fi"
+    ...
+  @
+-}
+data PathCompletionInfo = PathCompletionInfo
```

Comment:
The path completion stuff seems fairly generic, perhaps we could put it in its own module or something to emphasize that? I'm not sure we have anywhere else to actually _use_ it, but sometimes it's helpful just to know that something is conceptually independent.

Reply:
Yes, I think that is a good idea!